packages/graphiql-toolkit/src/create-fetcher/createFetcher.ts

```diff
@@ -52,6 +51,8 @@ export function createGraphiQLFetcher(options: CreateFetcherOptions): Fetcher {
       graphQLParams.operationName || undefined,
     );
     if (isSubscription) {
+      const wsFetcher = getWsFetcher(options, fetcherOpts);
```

Comment:
Looking at the types from `graphql-ws`, you can also pass a function as `connectionParams` to `createClient` which returns the payload for `connection_init`. Could we use this to avoid creating a new WS fetcher on each fetcher call?

Reply:
My concern with that approach was that when `headers` are changed and the subscription is re-issued, it might reuse the previous connection and end up with stale headers.

Explicitly restarting the connection every time seemed like the right thing to do and I specifically tested this scenario with this approach. 

I'm not sure if passing a promise ensures the same outcome. The connection might linger. 

I guess I could try it later on today if you think it's worth it.

Reply:
My only thought here was to avoid unnecessary work, but I doubt this adds much in terms of performance overhead. So would be fine for me to keep it like this if using a function doesn't work.

> I'm not sure if passing a promise ensures the same outcome.

I think the function can also be synchronous, at least it's typed like that:
```ts
connectionParams?: ConnectionInitMessage['payload'] | (() => Promise<ConnectionInitMessage['payload']> | ConnectionInitMessage['payload']);
```

Reply:
Yes, it can be asynchronous. But I assume that means it gets called just once when the connection is initiated.

The purpose of explicitly reconnecting is to allow modifying the connection options on the fly without refreshing the browser.

Reply:
I believe graphiql by default will reuse the websocket if the connection does not explicitly terminate.

Reply:
> The purpose of explicitly reconnecting is to allow modifying the connection options on the fly without refreshing the browser.

@andreialecu you can also change the fetcher prop in the GraphiQL component on the fly without a page reload, in a parent component state or otherwise, and for the next op execution of any kind, subscription or otherwise, it will use the newly configured params. 

If `schema` prop is not controlled by parent state, it will also re-load the schema with the changed params. Changing many of the props at runtime will update the GraphiQL component dynamically, because react. This is how many framework implementations manage this problem currently, with their own product wrapping the editor, and controlling schema and/or fetcher state with GraphiQL as a subcomponent. 

If we eventually get to our original 2019 proposal of having vscode style config strings ala playground, and can configure graphql-config like schema configs for projects, then we can make `fetcher` optional for once, and optionally create the fetcher internally, and solve this problem much more easily for users with static configs.

Reply:
> I believe graphiql by default will reuse the websocket if the connection does not explicitly terminate.

this is true, the connection init params are passed on initialising the websococket connection, so even though graphiql's state updates dynamically as `fetcher` props is updated at runtime, the connection needs to re-initialize to pass the new connection init params. I'm not sure if we would need this for all graphql subscription protocol implementations, so it's good that this happens within `createFetcher`